lib/spack/spack/util/archive.py

```diff
@@ -238,3 +241,42 @@ def reproducible_tarfile_from_prefix(
                 add_file(tar, file_info, entry.path)
 
         dir_stack.extend(reversed(new_dirs))  # we pop, so reverse to stay alphabetical
+
+
+def _git_prefix(archive_path):
+    # This is an annoying method, but since we always have a prefix and can't gaurantee what
+    # it is we need this.
+    tar = which("tar", required=True)
+    paths = tar("-tf", archive_path, output=str, error=str, fail_on_error=False)
+    if paths:
+        paths = paths.strip().split()
+        for p in paths:
+            if p.endswith(".git/"):
+                return p[:-5]
+    return ""
+
+
+def retrieve_commit_from_archive(archive_path, ref):
+    """extract git data from an archive with out expanding it"""
+    assert os.path.isfile(archive_path)
```

Comment:
Probably better to raise a FileNotFound error if this isn't true instead of an assertion.